### PR TITLE
Allow micro sign in identifiers

### LINKED ISSUE #11750: Compiler regression on 1.14 when using unicode as variables names
### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
```
Erlang/OTP 24 [erts-12.3] [source] [64-bit] [smp:32:32] [ds:32:32:10] [async-threads:1] [jit]

Elixir 1.14.0-dev (700e0b7) (compiled with Erlang/OTP 24)
```

* Operating system: 
```
Linux 5.17.1-arch1-1 x86_64
```

### Current behavior

Errors when compiling the current `tesla` release from hex (1.4.4).

```
== Compilation error in file lib/tesla/middleware/telemetry.ex ==
** (SyntaxError) lib/tesla/middleware/telemetry.ex:150:18: unexpected token: "µ" (column 18, code point U+00B5)
     |
 150 |         duration_µs = System.convert_time_unit(duration, :native, :microsecond)
     |                  ^
    (elixir 1.14.0-dev) lib/kernel/parallel_compiler.ex:346: anonymous fn/5 in Kernel.ParallelCompiler.spawn_workers/7
could not compile dependency :tesla, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile tesla", update it with "mix deps.update tesla" or clean it with "mix deps.clean tesla"
```

This code has since been removed from tesla but they haven't made a new release yet. It was addressed in [this PR](https://github.com/teamon/tesla/pull/519) and mentioned in [this issue](https://github.com/teamon/tesla/issues/524).

### Expected behavior

This code works on 1.13 and should continue working on 1.14.

### PR DESCRIPTION
**Pull request by josevalim**
Closes #11750.

/cc @mrluc

### CODE REVIEW

**Comment by mrluc** [comment]
Oh, cool -- I wondered if this would come up, since the original PR even had an example or two that needed removed from the Elixir test suite. I commented [on the issue](https://github.com/elixir-lang/elixir/issues/11750#issuecomment-1094157575) about a mechanism that Julia uses for this, 'canonicalization', that I think could also work for us if we wanted. It'd be kind of cool to use that mechanism together with the UTS39 protections, which is new-ish territory -- Julia AFAIR doesn't have those protections yet, but I can't think offhand of why they'd be incompatible as long as the canonicalizations are towards 'Common' scriptset characters, ie not switching to a script that could cause script-mixing issues.

**Comment by mrluc** [comment]
@josevalim I think we could merge this as a patch to fix the regression given how narrow the impact is, if we need time to fix it 'right'.

I have a candidate for the root cause now!

I now suspect the 'Common' scriptset and the definition of 'All' scriptset from UTS39 may have been casualties of the heroic UTS39C3 PR that refactored those protections into the tokenizer. Which would mean that I missed that when [reviewing](https://github.com/elixir-lang/elixir/pull/11621)

I think it was natural that tests 'from the outside' to continue to pass there and not catch that, as opposed to theoretical narrower ones against scriptset resolution which we didn't have, because until we were willing to start allowing Restricted codepoints, there'd be no inputs that could reveal that regression: ['Common'-scriptset characters that are also 'Allowed'](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=[%5b%3aScript%3dCommon%3a%5d%26%5b%3aIdentifier%5fStatus%3dAllowed%3a%5d]), and are also non-ASCII (0-9 and underscore), and also don't have a ScriptExtension classifying them into a writing system (😅), and are letters, don't actually exist!

So if the PR 🪓'd  the 'Common' scriptset/the technical definition of the 'All' set of script sets, it snuck by unnoticed -- until we had a reason to break one of the filters above, in this case to include a 'Restricted' codepoint.

I *think* we can reveal the issue by changing this PR's added mapping -- `scripts_special_cases` -- to contain `0x00B5 => ["Common"]` instead of 'Latin,' and you can see that it treats the character differently than the ones that are classified as '@top' of the scriptset bitmap, which I believe was intended to represent the definition of 'All' from UTS39 -- which is 'Common' | 'Inherited'.

Making that change results in the following, which treats underscore and micro differently; underscore is '@top' of the bitmap, micro is Common:

```
     Mixed-script identifiers are not supported for security reasons. 'foo_µs' is made of the following scripts:

       \u0066 f {Latin}
       \u006F o {Latin}
       \u006F o {Latin}
       \u005F _
       \u00B5 µ {Common}
       \u0073 s {Latin}
```

'All' scripts incl. 'Common', per UTS39 should be ignored for script resolution, like the underscore; that was done differently in the pre-refactored C3 PR.

- Aside: I think it's possible that the fall-through case for '@top' in `codepoint_to_scriptset` is actually UTS39's definition of 'unknown', ie it can only happen if there's no scriptset found in scripts/scriptextensions txt files. But it's protected from getting many inputs like that in practice, since it's getting codepoints that made it through all the other unicode filters.

File: lib/elixir/pages/unicode-syntax.md [hunk 1]
```diff
@@ -111,11 +111,11 @@ Elixir conforms to the clauses outlined in the [Unicode Technical Standard #39](
 
 ### C1. General Security Profile for Identifiers
 
-Elixir will not allow tokenization of identifiers with codepoints in `\p{Identifier_Status=Restricted}`.
+> An implementation following the General Security Profile does not permit any characters in \p{Identifier_Status=Restricted}, unless it documents the additional characters that it does allow
 
-> An implementation following the General Security Profile does not permit any characters in \p{Identifier_Status=Restricted}, ...
+Elixir will not allow tokenization of identifiers with codepoints in `\p{Identifier_Status=Restricted}`, with the single exception of MICRO SIGN µ (U+00B5), used by international system of units and assigned a scriptset of Latin.
```
Comment (mrluc) [inline lib/elixir/pages/unicode-syntax.md:116]:
```suggestion
Elixir will not allow tokenization of identifiers with codepoints in `\p{Identifier_Status=Restricted}`, with the single exception of MICRO SIGN µ (U+00B5), used by international system of units.
```

File: lib/elixir/test/elixir/kernel/warning_test.exs [hunk 2]
```diff
@@ -86,12 +89,15 @@ defmodule Kernel.WarningTest do
...
+      assert Code.eval_string("幻ㄒㄧㄤ = 1")
+      assert Code.eval_string("幻ㄒㄧㄤ1 = 1")
+      assert Code.eval_string("__सवव_1? = 1")
 
       # uts39 5.2 allowed 'highly restrictive' script mixing, like 't-shirt' in Jpan:
-      assert capture_err(fn -> Code.string_to_quoted!(":Tシャツ") end) == ""
+      assert Code.string_to_quoted!(":Tシャツ")
+
+      # allow Elixir special cases
+      assert Code.string_to_quoted!(":duration_µs")
```
Comment (mrluc) [inline lib/elixir/test/elixir/kernel/warning_test.exs:100]:
```suggestion
      assert Code.string_to_quoted!(":duration_µs")
      # Common 'micro' character can be used with any script
      assert Code.string_to_quoted!(":सवव_µ")
```

The

File: lib/elixir/unicode/tokenizer.ex [hunk 3]
```diff
@@ -1,6 +1,19 @@
...
+    # µ is removed because it normalizes to greek's lower mu.
+    # However, they can't be used at the same time as they
+    # belong to different scripts, so we add µ as latin.
+    #
+    # The mathematical characters between 1D6A8..1D7CB exhibit
+    # a similar property. However, because some of them can
+    # conflict with both Latin and Greek characters, such as 𝚳,
+    # they are not added here.
+    0x00B5 => ["Latin"]
+  }
```
Comment (mrluc) [inline lib/elixir/unicode/tokenizer.ex:15]:
```suggestion
  allowed_special_cases = [
    # µ is the Common-scriptset character for micro
    0x00B5
  ]
```

File: lib/elixir/unicode/tokenizer.ex [hunk 4]
```diff
@@ -86,8 +99,8 @@ defmodule String.Tokenizer do
       with [range, type_with_comments] <- :binary.split(line, ";"),
            [types, _comments] <- :binary.split(type_with_comments, "#"),
            types = String.split(types, " ", trim: true),
-           true <- "Inclusion" not in types and "Recommended" not in types do
-        range_to_codepoints.(range)
+           false <- "Inclusion" in types or "Recommended" in types do
+        Enum.reject(range_to_codepoints.(range), &Map.has_key?(scripts_special_cases, &1))
```
Comment (mrluc) [inline lib/elixir/unicode/tokenizer.ex:103]:
```suggestion
        Enum.reject(range_to_codepoints.(range), &(&1 in allowed_special_cases))
```

File: lib/elixir/unicode/tokenizer.ex [hunk 5]
```diff
@@ -151,7 +164,7 @@ defmodule String.Tokenizer do
 
   scripts = codepoints_to_scriptset.("Scripts.txt", %{})
   script_extensions = codepoints_to_scriptset.("ScriptExtensions.txt", script_aliases)
-  all_codepoints_to_scriptset = scripts ++ script_extensions
+  all_codepoints_to_scriptset = Map.to_list(scripts_special_cases) ++ scripts ++ script_extensions
```
Comment (mrluc) [inline lib/elixir/unicode/tokenizer.ex:167]:
```suggestion
  all_codepoints_to_scriptset = scripts ++ script_extensions
```

**Review by mrluc — COMMENTED** [review]
In local testing when pulling this branch, it seems like the only tweak needed is to whitelist the Restricted codepoint; I tested with `make clean test elixir`; I tried to add suggested changes through the GH UI that match.

That done, I think all of the other logic is actually good -- including your logic that lets the 'Common' scriptset 'fall through' to `@top`, @josevalim; that was a red herring on my part, due to the way I was testing it.

**Comment by josevalim** [comment]
The issue with allowing micro everywhere, is that someone would be able to mix it with Greek. We could give it everything but Greek. Or convert the Greek lowercase mu to micro, which is also an approach I like.

I found the approach of automatically performing normalization interesting, however I would definitely stick with NFC. I would not be comfortable with having different looking Characters being the same because it can introduce the same kind of confusion, which would definitely happen with NFKC.  Could this happen with NFC?

I think we also need to improve the error message for identifiers with NFKC characters. Today it just says it is invalid and perhaps we should say the real reason, but this depends on our choices above.

**Comment by mrluc** [comment]
> The issue with allowing micro everywhere, is that someone would be able to mix it with Greek

True; technically nearly the same is true of the Cyrillic/Greek etc chars that are visually identical to the latin ones, and are currently allowed. But it's definitely true that the Common micro should be identical to the Greek mu whereas IDE sometimes try to show scriptset differences, like cycling fonts  -- anecdata, in Emacs I notice word-jump starts a new word when the scriptset switches to cyrillic, and not when using the Common micro character, which is cool! It cycles through a few monotype fonts as well for Cyrillic/Greek/Latin too, though that's faint.

Thinking about the failure experience in both cases: "some code uses one µ character, and then someone else tries to type the same variable with the wrong µ character, and it doesn't work right!" -- but that one should actually get a helpful message, thanks to confusability warnings.

> convert the Greek lowercase mu to micro, which is also an approach I like.

Nice, yeah that seems like a good experience, canonicalization a la Julia. They do it at/around the same time as normalization I think, via the extension to that utf8proc library, but it's conceptually a separate thing and could be simple to implement in the tokenizer.

> I think we also need to improve the error message for identifiers with NFKC characters. Today it just says it is invalid and perhaps we should say the real reason, but this depends on our choices above.

Maybe ideally, it could direct them to the unicode syntax page in the elixir docs or a certain section of it, where it defines what's valid in an identifier; I wouldn't say we'd need to mention normalization specifically -- Unicode has Not_NFKC as the kind of 'Restricted' it is, but really the reason (from our point of view) is just that Unicode defines it as a Restricted character, and it's one that Elixir hasn't yet made an exception for.

**Comment by josevalim** [comment]
I think I have a more concrete idea of how we can move forward. We can say we perform NFC normalization by default and some NFKC normalization. At first, the only NFKC normalization we will perform is micro to mu. For the normalization, We will change The script set To be The union of the script set of all involved characters, which in this case, will be top.

For the error messages, if we have an invalid token, we can apply NFKC and try to parse it again in elixir_tokenizer. If it succeEds,  the error message can tell the user how to properly write the identifier.

we can implement this in three distinct PR’s:

1. apply NFC automatically instead of raising
2. Apply some NFKC as explained above
3. Improve error messages for NFKC 

What do you think? Would you like to send PR’s to some of those? :)

**Comment by mrluc** [comment]
Oh, cool! Sounds good, yeah I didn't have clear the advantage of going 'Common' => 'Greek', and thus why we would want to support changing scripts in the overrides/exceptions at all -- but MICRO decomposing to MU in NFKC is a decent reason, and the extra control it gives as in the Latin->Greek vs. separate Ukrainian.

So it sounds fun and I think I could use some shop time on it -- the only thing is that I *may* not be able to start until this coming Thursday, have a busy Mon-Wed this week (flying back from south america and doing some client work Mon-Wed, rails + laravel stuff). I *might* be able to get into it earlier (big layover, heh) -- so if someone else wants to grab it of course, fire away.

**Comment by josevalim** [comment]
Perfect. take your time! I also have a wrist injurty, so i can't do much. :) I have opened up another issue so we can track this.

Closing in favor of #11753.

**Pull request closed without merging**